ARMv7 segmentation fault in oracle.readTs when calling loadUint64

I am facing an issue running badger on an ARMv7 architecture. The minimal test case below works quite fine on an amd64 machine but, unfortunately, not on ARMv7 32bit.

The trace below shows that the issue originates in ``atomic.loadUint64()`` but I also run basic atomic operations tests against the golang runtime, and they work fine on this architecture. 

It looks to me that the underlying memory of ``oracle.curRead`` somehow vanishes but I am not sure.

Below you find also a strace trace. There the segmentation fault happens after the ``madvise``, but I am not sure if this is related. 

Badger version: 1.0.1 (89689ef36cae26ae094cb5ea79b7400d839f2d68)
golang version: 1.8.5 and 1.9.2

Test case:
```golang
func TestPersistentCache_DirectBadger(t *testing.T) {
	dir, err := ioutil.TempDir("", "")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(dir)

	config := badger.DefaultOptions
	config.TableLoadingMode = options.MemoryMap
	config.ValueLogFileSize = 16 << 20
	config.LevelOneSize = 8 << 20
	config.MaxTableSize = 2 << 20
	config.Dir = dir
	config.ValueDir = dir
	config.SyncWrites = false

	db, err := badger.Open(config)
	if err != nil {
		t.Fatalf("cannot open db at location %s: %v", dir, err)
	}

	err = db.View(func(txn *badger.Txn) error { return nil })

	if err != nil {
		t.Fatal(err)
	}

	if err = db.Close(); err != nil {
		t.Fatal(err)
	}
}
```

```golang
=== RUN   TestPersistentCache_DirectBadger
--- FAIL: TestPersistentCache_DirectBadger (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x1150c]

goroutine 5 [running]:
testing.tRunner.func1(0x10a793b0)
        /usr/lib/go/src/testing/testing.go:711 +0x2a0
panic(0x3e4bd8, 0x6bb478)
        /usr/lib/go/src/runtime/panic.go:491 +0x204
sync/atomic.loadUint64(0x10a483cc, 0x200000, 0x0)
        /usr/lib/go/src/sync/atomic/64bit_arm.go:10 +0x3c
github.com/grid-x/client/vendor/github.com/dgraph-io/badger.(*oracle).readTs(0x10a483c0, 0x14, 0x5)
        /home/robert/Projects/gridx/client/src/github.com/grid-x/client/vendor/github.com/dgraph-io/badger/transaction.go:87 +0x3c
github.com/grid-x/client/vendor/github.com/dgraph-io/badger.(*DB).NewTransaction(0x10b06000, 0x0, 0x4cccc)
        /home/robert/Projects/gridx/client/src/github.com/grid-x/client/vendor/github.com/dgraph-io/badger/transaction.go:440 +0x20
github.com/grid-x/client/vendor/github.com/dgraph-io/badger.(*DB).View(0x10b06000, 0x464e20, 0x0, 0x0)
        /home/robert/Projects/gridx/client/src/github.com/grid-x/client/vendor/github.com/dgraph-io/badger/transaction.go:457 +0x3c
command-line-arguments.TestPersistentCache_DirectBadger(0x10a793b0)
        /home/robert/Projects/gridx/client/src/github.com/grid-x/client/pkg/cache/persistent_cache_test.go:64 +0x1e8
testing.tRunner(0x10a793b0, 0x464e24)
        /usr/lib/go/src/testing/testing.go:746 +0xb0
created by testing.(*T).Run
        /usr/lib/go/src/testing/testing.go:789 +0x258
```

strace:
```strace
[pid 15075] mmap2(NULL, 33554432, PROT_READ, MAP_SHARED, 6, 0) = 0xb4dff000                                   
[pid 15075] madvise(0xb4dff000, 33554432, MADV_RANDOM) = 0                     
[pid 15075] clock_gettime(CLOCK_MONOTONIC, {tv_sec=69709, tv_nsec=217038306}) = 0                            
[pid 15075] --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x4} ---
[pid 15075] rt_sigreturn()              = 0                                       
```

Here a clean build from your https://github.com/dgraph-io/badger/tree/issue-311 branch. I will attach a full strace in a couple of minutes

```golang
root@scw-546b0d:~# ./badger_tests.arm -test.v -test.run ^TestPersistentCache_DirectBadger$
=== RUN   TestPersistentCache_DirectBadger
All good. about to load o.curRead
--- FAIL: TestPersistentCache_DirectBadger (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x1150c]

goroutine 5 [running]:
testing.tRunner.func1(0x10896c60)
        /usr/lib/go/src/testing/testing.go:711 +0x2a0
panic(0x35f358, 0x58dff0)
        /usr/lib/go/src/runtime/panic.go:491 +0x204
sync/atomic.loadUint64(0x1084a84c, 0x22, 0x0)
        /usr/lib/go/src/sync/atomic/64bit_arm.go:10 +0x3c
github.com/dgraph-io/badger.(*oracle).readTs(0x1084a840, 0x14, 0x5)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/transaction.go:88 +0x68
github.com/dgraph-io/badger.(*DB).NewTransaction(0x10b0e000, 0x0, 0x4cccc)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/transaction.go:443 +0x20
github.com/dgraph-io/badger.(*DB).View(0x10b0e000, 0x3c14e8, 0x0, 0x0)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/transaction.go:460 +0x3c
github.com/dgraph-io/badger.TestPersistentCache_DirectBadger(0x10896c60)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/issue311_test.go:32 +0x1e8
testing.tRunner(0x10896c60, 0x3c14ec)
        /usr/lib/go/src/testing/testing.go:746 +0xb0
created by testing.(*T).Run
        /usr/lib/go/src/testing/testing.go:789 +0x258
```

Ok I can replicate the failure on the Baremetal C1. However, I am not sure if this is something to do with Badger. Please check the latest code I pushed to the branch. 

This test which does not exercise any Badger code, fails as well:

```Go
type orc struct {
	isManaged bool // Does not change value, so no locking required.

	sync.Mutex
	curRead    uint64
	nextCommit uint64

	// These two structures are used to figure out when a commit is done. The minimum done commit is
	// used to update curRead.
	commitMark     uint64Heap
	pendingCommits map[uint64]struct{}

	// commits stores a key fingerprint and latest commit counter for it.
	// refCount is used to clear out commits map to avoid a memory blowup.
	commits  map[uint64]uint64
	refCount int64
}

func TestAtomicLoadArm(t *testing.T) {
	o := &orc{}
	atomic.LoadUint64(&o.curRead)
}
```
Test output:

```
 go test -v  -run ^TestAtomic
=== RUN   TestAtomicLoadArm
--- FAIL: TestAtomicLoadArm (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x11d64]

goroutine 5 [running]:
testing.tRunner.func1(0x10812ab0)
        /usr/lib/go-1.9/src/testing/testing.go:711 +0x2a0
panic(0x35fb70, 0x58e160)
        /usr/lib/go-1.9/src/runtime/panic.go:491 +0x204
sync/atomic.loadUint64(0x1085484c, 0x10854840, 0xe9630)
        /usr/lib/go-1.9/src/sync/atomic/64bit_arm.go:10 +0x3c
github.com/dgraph-io/badger.TestAtomicLoadArm(0x10812ab0)
        /root/go/src/github.com/dgraph-io/badger/issue311_test.go:61 +0x2c
testing.tRunner(0x10812ab0, 0x3c20ac)
        /usr/lib/go-1.9/src/testing/testing.go:746 +0xb0
created by testing.(*T).Run
        /usr/lib/go-1.9/src/testing/testing.go:789 +0x258
exit status 2
FAIL    github.com/dgraph-io/badger     0.071s
```
I would suspect there is an issue with the Go runtime, or the kernel which is causing this. I don’t think we can do much here. Maybe file a bug upstream?

:thinking: ``value`` is the first field in ``node`` but it fails with the same panic. This happens if I extend your test with a write transaction.

```patch
diff --git a/transaction_test.go b/transaction_test.go
index c234f65..5b2a9b6 100644
--- a/transaction_test.go
+++ b/transaction_test.go
@@ -641,7 +641,13 @@ func TestArmV7Issue311Fix(t *testing.T) {
        }
 
        err = db.View(func(txn *Txn) error { return nil })
+       if err != nil {
+               t.Fatal(err)
+       }
 
+       err = db.Update(func(txn *Txn) error {
+               return txn.Set([]byte{ 0x11 }, []byte{ 0x22 })
+       })
        if err != nil {
                t.Fatal(err)
        }
```

```tace
root@gridBox-B313-200-000-000-021-P-X:/data# ./badger_tests.arm -test.v -test.run ^TestArmV7Issue311Fix$
=== RUN   TestArmV7Issue311Fix
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x1150c]

goroutine 10 [running]:
sync/atomic.loadUint64(0x10af008c, 0xffff, 0x10af007e)
        /usr/lib/go/src/sync/atomic/64bit_arm.go:10 +0x3c
github.com/dgraph-io/badger/skl.(*node).getValueOffset(0x10af008c, 0x108561e8, 0x2ce8a0)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/skl/skl.go:138 +0x20
github.com/dgraph-io/badger/skl.(*Iterator).Value(0x1080c808, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/skl/skl.go:425 +0x2c
github.com/dgraph-io/badger.writeLevel0Table(0x108a1520, 0x1080c7f8, 0x0, 0x0)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/db.go:735 +0xec
github.com/dgraph-io/badger.(*DB).flushMemtable(0x10aee000, 0x1080f0c0, 0x0, 0x0)
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/db.go:778 +0x1cc
created by github.com/dgraph-io/badger.Open
        /home/robert/Projects/gridx/badger/src/github.com/dgraph-io/badger/db.go:257 +0x928
```

--- a/db_test.go
+++ b/db_test.go
@@ -328,7 +328,7 @@ func TestGetMore(t *testing.T) {
 	}
 	//	n := 500000
 	n := 10000
-	m := 49 // Increasing would cause ErrTxnTooBig
+	m := 45 // Increasing would cause ErrTxnTooBig
 	for i := 0; i < n; i += m {
 		txn := kv.NewTransaction(true)
 		for j := i; j < i+m && j < n; j++ {
@@ -432,7 +432,7 @@ func TestExistsMore(t *testing.T) {
 
 	//	n := 500000
 	n := 10000
-	m := 49
+	m := 45
 	for i := 0; i < n; i += m {
 		if (i % 1000) == 0 {
 			t.Logf("Putting i=%d\n", i)

--- a/levels.go
+++ b/levels.go
@@ -32,13 +32,12 @@ import (
 )
 
 type levelsController struct {
-	elog trace.EventLog
+	nextFileID uint64 // Atomic
+	elog       trace.EventLog
 
 	// The following are initialized once and const.
 	levels []*levelHandler
 	kv     *DB
-
-	nextFileID uint64 // Atomic
 
 	cstatus compactStatus
 }

--- a/skl/skl.go
+++ b/skl/skl.go
@@ -50,18 +50,19 @@ const (
 const MaxNodeSize = int(unsafe.Sizeof(node{}))
 
 type node struct {
+	// Multiple parts of the value are encoded as a single uint64 so that it
+	// can be atomically loaded and stored:
+	//   value offset: uint32 (bits 0-31)
+	//   value size  : uint16 (bits 32-47)
+	// 12 bytes are allocated to ensure 8 byte alignment also on 32bit systems.
+	value [12]byte
+
 	// A byte slice is 24 bytes. We are trying to save space here.
 	keyOffset uint32 // Immutable. No need to lock to access key.
 	keySize   uint16 // Immutable. No need to lock to access key.
 
 	// Height of the tower.
 	height uint16
-
-	// Multiple parts of the value are encoded as a single uint64 so that it
-	// can be atomically loaded and stored:
-	//   value offset: uint32 (bits 0-31)
-	//   value size  : uint16 (bits 32-47)
-	value uint64
 
 	// Most nodes do not need to use the full height of the tower, since the
 	// probability of each successive level decreases exponentially. Because
@@ -108,7 +109,7 @@ func newNode(arena *Arena, key []byte, v y.ValueStruct, height int) *node {
 	node.keyOffset = arena.putKey(key)
 	node.keySize = uint16(len(key))
 	node.height = uint16(height)
-	node.value = encodeValue(arena.putVal(v), v.EncodedSize())
+	*node.value64BitAlignedPtr() = encodeValue(arena.putVal(v), v.EncodedSize())
 	return node
 }
 
@@ -134,8 +135,15 @@ func NewSkiplist(arenaSize int64) *Skiplist {
 	}
 }
 
+func (s *node) value64BitAlignedPtr() *uint64 {
+	if uintptr(unsafe.Pointer(&s.value))%8 == 0 {
+		return (*uint64)(unsafe.Pointer(&s.value))
+	}
+	return (*uint64)(unsafe.Pointer(&s.value[4]))
+}
+
 func (s *node) getValueOffset() (uint32, uint16) {
-	value := atomic.LoadUint64(&s.value)
+	value := atomic.LoadUint64(s.value64BitAlignedPtr())
 	return decodeValue(value)
 }
 
@@ -146,7 +154,7 @@ func (s *node) key(arena *Arena) []byte {
 func (s *node) setValue(arena *Arena, v y.ValueStruct) {
 	valOffset := arena.putVal(v)
 	value := encodeValue(valOffset, v.EncodedSize())
-	atomic.StoreUint64(&s.value, value)
+	atomic.StoreUint64(s.value64BitAlignedPtr(), value)
 }
 
 func (s *node) getNextOffset(h int) uint32 {

--- a/transaction.go
+++ b/transaction.go
@@ -46,10 +46,11 @@ func (u *uint64Heap) Pop() interface{} {
 }
 
 type oracle struct {
+	curRead   uint64 // Managed by the mutex.
+	refCount  int64
 	isManaged bool // Does not change value, so no locking required.
 
 	sync.Mutex
-	curRead    uint64
 	nextCommit uint64
 
 	// These two structures are used to figure out when a commit is done. The minimum done commit is
@@ -59,8 +60,7 @@ type oracle struct {
 
 	// commits stores a key fingerprint and latest commit counter for it.
 	// refCount is used to clear out commits map to avoid a memory blowup.
-	commits  map[uint64]uint64
-	refCount int64
+	commits map[uint64]uint64
 }
 
 func (o *oracle) addRef() {
@@ -192,7 +192,7 @@ type Txn struct {
 func (txn *Txn) checkSize(e *entry) error {
 	count := txn.count + 1
 	// Extra bytes for version in key.
-	size := txn.size + int64(e.estimateSize(txn.db.opt.ValueThreshold)) + 10 
+	size := txn.size + int64(e.estimateSize(txn.db.opt.ValueThreshold)) + 10
 	if count >= txn.db.opt.maxBatchCount || size >= txn.db.opt.maxBatchSize {
 		return ErrTxnTooBig
 	}

--- a/transaction_test.go
+++ b/transaction_test.go
@@ -23,6 +23,7 @@ import (
 	"strconv"
 	"testing"
 
+	"github.com/dgraph-io/badger/options"
 	"github.com/dgraph-io/badger/y"
 
 	"github.com/stretchr/testify/require"
@@ -616,4 +617,50 @@ func TestManagedDB(t *testing.T) {
 		}
 	}
 	txn.Discard()
+}
+
+func TestArmV7Issue311Fix(t *testing.T) {
+	dir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	config := DefaultOptions
+	config.TableLoadingMode = options.MemoryMap
+	config.ValueLogFileSize = 16 << 20
+	config.LevelOneSize = 8 << 20
+	config.MaxTableSize = 2 << 20
+	config.Dir = dir
+	config.ValueDir = dir
+	config.SyncWrites = false
+
+	db, err := Open(config)
+	if err != nil {
+		t.Fatalf("cannot open db at location %s: %v", dir, err)
+	}
+
+	err = db.View(func(txn *Txn) error { return nil })
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = db.Update(func(txn *Txn) error {
+		return txn.Set([]byte{0x11}, []byte{0x22})
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = db.Update(func(txn *Txn) error {
+		return txn.Set([]byte{0x11}, []byte{0x22})
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = db.Close(); err != nil {
+		t.Fatal(err)
+	}
 }

